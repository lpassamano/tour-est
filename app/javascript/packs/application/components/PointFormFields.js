import React, { Component } from "react";
import PropTypes from "prop-types";
import { Button } from "react-foundation";
import { omit } from "lodash";
import ImageInput from "./ImageInput";
import { Link } from "@reach/router";

const INITIAL_STATE = {
  title: "",
  caption: "",
  description: "",
  location: "",
  directions: "",
  image: undefined,
  imageEdited: false
};

export class PointFormFields extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ...INITIAL_STATE,
      ...this.props.initialValues
    };
  }

  handleSubmit = event => {
    event.preventDefault();

    if (this.state.imageEdited) {
      this.props.onSubmit(omit(this.state, "imageEdited"));
    } else {
      this.props.onSubmit(omit(this.state, ["image", "imageEdited"]));
    }

    if (event.target.type === "button") {
      return this.setState(INITIAL_STATE);
    }

    this.props.onHide();
  };

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleChangeImage = image => {
    this.setState({
      image: image,
      imageEdited: true
    });
  };

  handleDeleteImage = event => {
    event.preventDefault();
    this.setState({
      image: undefined,
      imageEdited: true
    });
  };

  render() {
    const {
      title,
      caption,
      description,
      location,
      directions,
      image
    } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <label htmlFor="image">Upload Image</label>
        <ImageInput
          name="image"
          onChange={this.handleChangeImage}
          value={image}
        />
        {image && (
          <Link to="#" onClick={this.handleDeleteImage} id="delete-image">
            <i className="fi-trash" /> Remove Image
          </Link>
        )}
        <label htmlFor="title">Title: </label>
        <input
          name="title"
          type="text"
          value={title}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="caption">Caption: </label>
        <textarea
          name="caption"
          type="text"
          value={caption}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="description">Description: </label>
        <textarea
          name="description"
          type="text"
          value={description}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="location">Location: </label>
        <input
          name="location"
          type="text"
          value={location}
          onChange={this.handleChange}
        />
        <br />
        <label htmlFor="directions">Directions: </label>
        <textarea
          name="directions"
          type="text"
          value={directions}
          onChange={this.handleChange}
        />
        <br />
        <Button type="submit">Save</Button>
        <Button type="button" onClick={this.handleSubmit}>
          Save and add another
        </Button>
        <br />
        <Button className="hollow" type="button" onClick={this.props.onHide}>
          Cancel
        </Button>
      </form>
    );
  }
}

PointFormFields.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onHide: PropTypes.func.isRequired,
  initialValues: PropTypes.shape({
    title: PropTypes.string,
    caption: PropTypes.string,
    description: PropTypes.string,
    location: PropTypes.string,
    directions: PropTypes.string,
    image: PropTypes.string
  })
};

PointFormFields.defaultProps = {
  initialValues: {}
};

export default PointFormFields;
