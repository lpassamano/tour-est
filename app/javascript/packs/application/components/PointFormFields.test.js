import { PointFormFields } from "./PointFormFields";
import React from "react";
import { shallow } from "enzyme";

describe("<PointFormFields />", () => {
  const setup = ({ onSubmit = jest.fn() }) => {
    const onHide = jest.fn();
    const component = shallow(
      <PointFormFields onSubmit={onSubmit} onHide={onHide} />
    );
    return component;
  };

  const fillIn = (input, value) => {
    input.simulate("change", {
      target: { value, name: input.prop("name") }
    });
  };

  const submit = form => {
    const mockedEvent = { preventDefault: jest.fn(), target: {} };
    form.simulate("submit", mockedEvent);
  };

  test("handleChange - when text is typed into the fields onChange is called", () => {
    const component = setup({});
    fillIn(component.find('[name="title"]'), "Nike of Samothrace");
    expect(component.state("title")).toEqual("Nike of Samothrace");
  });

  test("handleChangeImage - when a new image is uploaded handleChangeImage is called", () => {
    const component = setup({});
    component.find('[name="image"]').simulate("change", {
      preview: "blob:http://localhost:300/a3b6c1fa-93b9",
      name: "nike.jpeg",
      size: 197764,
      type: "image/jpeg"
    });

    expect(component.state("imageEdited")).toEqual(true);
    expect(component.state("image")).toEqual({
      preview: "blob:http://localhost:300/a3b6c1fa-93b9",
      name: "nike.jpeg",
      size: 197764,
      type: "image/jpeg"
    });
  });

  test("onSubmit - when button is clicked the current state is submitted without an image", () => {
    const onSubmit = jest.fn();
    onSubmit.mockResolvedValue({ ok: true, data: { id: 1 } });
    const component = setup({ onSubmit });
    fillIn(component.find('[name="title"]'), "Nike of Samothrace");
    submit(component.find("form"));

    expect(onSubmit).toHaveBeenCalledWith({
      title: "Nike of Samothrace",
      caption: "",
      description: "",
      directions: "",
      location: ""
    });
  });

  test("onSubmit - when button is clicked the current state is submitted with an image", () => {
    const onSubmit = jest.fn();
    onSubmit.mockResolvedValue({ ok: true, data: { id: 1 } });
    const component = setup({ onSubmit });
    fillIn(component.find('[name="title"]'), "Nike of Samothrace");
    component.find('[name="image"]').simulate("change", {
      preview: "blob:http://localhost:300/a3b6c1fa-93b9",
      name: "nike.jpeg",
      size: 197764,
      type: "image/jpeg"
    });
    submit(component.find("form"));

    expect(onSubmit).toHaveBeenCalledWith({
      title: "Nike of Samothrace",
      caption: "",
      description: "",
      directions: "",
      location: "",
      image: {
        preview: "blob:http://localhost:300/a3b6c1fa-93b9",
        name: "nike.jpeg",
        size: 197764,
        type: "image/jpeg"
      }
    });
  });

  test("render", () => {
    const component = setup({});
    expect(component).toMatchSnapshot();
  });
});
