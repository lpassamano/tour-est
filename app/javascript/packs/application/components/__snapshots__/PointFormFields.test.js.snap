// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PointFormFields /> render 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    htmlFor="image"
  >
    Upload Image
  </label>
  <ImageInput
    name="image"
    onChange={[Function]}
  />
  <label
    htmlFor="title"
  >
    Title: 
  </label>
  <input
    name="title"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="caption"
  >
    Caption: 
  </label>
  <textarea
    name="caption"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="description"
  >
    Description: 
  </label>
  <textarea
    name="description"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="location"
  >
    Location: 
  </label>
  <input
    name="location"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <label
    htmlFor="directions"
  >
    Directions: 
  </label>
  <textarea
    name="directions"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
  <Button
    type="submit"
  >
    Save
  </Button>
  <Button
    onClick={[Function]}
    type="button"
  >
    Save and add another
  </Button>
  <br />
  <Button
    className="hollow"
    onClick={[MockFunction]}
    type="button"
  >
    Cancel
  </Button>
</form>
`;
